### PR TITLE
Add constructor / equality checks for dns, email and uri types

### PR DESCRIPTION
**Pull request by ozevren**

### CODE REVIEW

File: mixer/pkg/il/runtime/externs.go [hunk 1]
```diff
@@ -111,6 +136,158 @@ func externTimestampEqual(t1 time.Time, t2 time.Time) bool {
...
+	}
+
+	return local1 == local2, nil
+}
+
+func externUri(in string) (string, error) {
+	if in == "" {
+		return "", errors.New("error converting string to url: empty string")
+	}
+	_, err := url.Parse(in)
```
Comment (geeknoid) [inline mixer/pkg/il/runtime/externs.go]: Combine with next line.

File: mixer/pkg/il/runtime/externs.go [hunk 2]
```diff
@@ -111,6 +136,158 @@ func externTimestampEqual(t1 time.Time, t2 time.Time) bool {
...
+	if err != nil {
+		return "", fmt.Errorf("error converting string to url '%s': '%v'", in, err)
+	}
+	return in, err
```
Comment (geeknoid) [inline mixer/pkg/il/runtime/externs.go]: return in, nil

File: mixer/pkg/il/runtime/externs.go [hunk 3]
```diff
@@ -111,6 +136,158 @@ func externTimestampEqual(t1 time.Time, t2 time.Time) bool {
...
+		return false, fmt.Errorf("error converting string to url '%s': '%v'", u1, err)
+	}
+
+	url2, err := url.Parse(u2)
+	if err != nil {
+		return false, fmt.Errorf("error converting string to url '%s': '%v'", u2, err)
+	}
+
+	// Try to apply as much normalization logic as possible.
+	if url1.Scheme != url2.Scheme {
```
Comment (geeknoid) [inline mixer/pkg/il/runtime/externs.go]: SHould this be a case-insensitive check?

File: mixer/pkg/il/runtime/externs.go [hunk 4]
```diff
@@ -111,6 +136,158 @@ func externTimestampEqual(t1 time.Time, t2 time.Time) bool {
...
+		return false, nil
+	}
+
+	if url1.Scheme == "http" || url1.Scheme == "https" {
```
Comment (geeknoid) [inline mixer/pkg/il/runtime/externs.go]: HTTP and HTTPS?

File: mixer/pkg/il/runtime/externs.go [hunk 5]
```diff
@@ -111,6 +136,158 @@ func externTimestampEqual(t1 time.Time, t2 time.Time) bool {
...
+
+		if url1.Port() != url2.Port() {
+			return false, nil
+		}
+
+		// normalize url2 to url1's hostname.
+		url1.Host = url2.Host
+	}
+
+	return url1.String() == url2.String(), nil
```
Comment (geeknoid) [inline mixer/pkg/il/runtime/externs.go]: Case insensitive?

File: mixer/pkg/il/runtime/externs.go (hunk 3 above)
Comment (ozevren) [reply mixer/pkg/il/runtime/externs.go]: Great catch. Thanks. Added support.

File: mixer/pkg/il/runtime/externs.go (hunk 5 above)
Comment (ozevren) [reply mixer/pkg/il/runtime/externs.go]: The case sensitive parts are (for http/https) are normalized within the if block. Other parts are case sensitive.

File: mixer/pkg/il/runtime/externs.go [hunk 6]
```diff
@@ -111,6 +137,165 @@ func externTimestampEqual(t1 time.Time, t2 time.Time) bool {
...
+		return "", fmt.Errorf("error converting '%s' to e-mail: display names are not allowed", in)
+	}
+
+	// Also check through the dns name logic to ensure that this will not cause any breaks there, when used for
+	// comparison.
+
+	_, domain := getEmailParts(a.Address)
+
+	_, err = externDNSName(domain)
+	if err != nil {
```
Comment (geeknoid) [inline mixer/pkg/il/runtime/externs.go]: Combine with previous line

**Pull request merged**